### ISSUE TITLE
crash: Updated constraint could not find existing matching constraint to update

### ISSUE THREAD

**Issue by tcurdt** [SnapKit/SnapKit#295]
### New Issue Checklist

* [x] I have looked at the [Documentation](http://snapkit.io/docs)
* [x] I have read the [F.A.Q.](http://snapkit.io/faq)

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform                | ios
 Platform Version        | 10
 SnapKit Version         | 0.40.0 (2e4cebf41d8bd0833487381590ed90c38b9a2491)
 Integration Method      | carthage
 

### Issue Description
I've just migrated my app to Swift3 and hence also switched to version 0.40 unfortunately it now crashes on startup. I am seeing

```
NSHashTable {
[0] <SnapKit.LayoutConstraint:0x6000000bf3e0@ReloadView.swift#47 UILabel:0x7faa7bc38770.top == Foo.ReloadView:0x7faa7bc35340.top + 20.0>
}
 NSHashTable {
[0] <SnapKit.LayoutConstraint:0x6000000bf3e0@ReloadView.swift#47 UILabel:0x7faa7bc38770.top == Foo.ReloadView:0x7faa7bc35340.top + 20.0>
}

fatal error: Updated constraint could not find existing matching constraint to update: <SnapKit.LayoutConstraint:0x6000000bf440@ReloadView.swift#48 UILabel:0x7faa7bc38770.left == Foo.ReloadView:0x7faa7bc35340.left + 20.0>: file /project/Carthage/Checkouts/SnapKit/Source/Constraint.swift, line 233
```
 
A stripped down version of the view controller and view looks like this:

```
class ReloadView: UIView {

    override init(frame: CGRect) {
        super.init(frame: frame)
        configure()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        configure()
    }

    convenience init() {
        self.init(frame: CGRect.zero)
    }

    func configure() {

        let label = UILabel()
        let button = UIButton(type: UIButtonType.roundedRect)

        addSubview(label)
        addSubview(button)

        let padding = 20

        label.snp.updateConstraints { make in
            make.top.equalTo(padding)
            make.left.equalTo(self).offset(padding)
            make.right.equalTo(self).offset(-padding)
            make.width.equalTo(300)
        }

        button.snp.updateConstraints { make in
            make.top.equalTo(label.snp.bottom).offset(padding)
            make.left.equalTo(padding)
            make.width.equalTo(self).offset(-2*padding)
            make.bottom.equalTo(0).offset(-padding)
        }
    }
}
```

```
class FooViewController: UIViewController {

    private var reloadView: ReloadView!

    override func viewDidLoad() {
        super.viewDidLoad()

        let reloadView = ReloadView()
        self.reloadView = reloadView

        view.addSubview(reloadView)

        reloadView.snp.updateConstraints { make in
            // make.center.equalToSuperview()
            make.edges.equalTo(view.snp.edges)
        }

    }
}
```

**Comment by tcurdt** [comment]
Commenting out that view I am off to the next problem:

```
*** Terminating app due to uncaught exception 'NSInvalidLayoutConstraintException', reason: 'Constraint improperly relates anchors of incompatible types: <SnapKit.LayoutConstraint:0x6000000b9c80@WebViewController.swift#84 NSKVONotifying_WKWebView:0x7fb8b482ce00.top == UIView:0x7fb8b3d0ca30.left>'
```
with the constraints being

```
        webView.snp.updateConstraints { make in
            // make.edges.equalToSuperview()
            // make.edges.equalTo(view)
            make.edges.equalTo(view.snp.edges)
        }
```

**Comment by tcurdt** [comment]
Seems like `make.top.left.bottom.right.equalTo(view)` does work.

**Comment by robertjpayne** [comment]
@tcurdt I think this is a dupe of #291

**Comment by robertjpayne** [comment]
@tcurdt yea this is a dupe of #291, the crasher is due to a new way `updateConstraints` works that does not allow you to create *new* constraints if there is already one or more existing constraints from SnapKit on the view.

**Comment by tcurdt** [comment]
@robertjpayne I've run into another thing that seems related:

```
targetView.snp.updateConstraints { make in
    make.edges.equalToSuperview().inset(UIEdgeInsets(top: 0, left: 0, bottom: 150, right: 0))
}

let margin = UIEdgeInsets(top: 8, left: 8, bottom: -8, right: -8)

cornerTL.snp.updateConstraints { make in // an UIImageView
    make.top.equalTo(targetView).offset(margin.top)
    make.left.equalTo(targetView).offset(margin.left)
}
```

results in

```
fatal error: Updated constraint could not find existing matching constraint to update: <SnapKit.LayoutConstraint:0x6080002acfc0@ScanController.swift#184 UIImageView:0x7fe923426270.left == UIView:0x7fe923649780.left + 8.0>: file /project/Carthage/Checkouts/SnapKit/Source/Constraint.swift, line 233

```

**Comment by robertjpayne** [comment]
@tcurdt the error message it accurate, you're attempting to update constraints that do not exist likely (is targetView the same as the original superview?)

**Comment by robertjpayne** [comment]
@tcurdt wait, this isn't quite right, do you have *any* other code blocks that do `cornerTL.snp.updateConstraints`?

**Comment by tcurdt** [comment]
to give the bigger picture:

```
let targetView = UIView()
view.addSubview(targetView)

let cornerTL = UIImageView(image: ...)
targetView.addSubview(cornerTL)

targetView.snp.updateConstraints { make in
    make.edges.equalToSuperview().inset(UIEdgeInsets(top: 0, left: 0, bottom: 150, right: 0))
}

let margin = UIEdgeInsets(top: 8, left: 8, bottom: -8, right: -8)

cornerTL.snp.updateConstraints { make in // an UIImageView
    make.top.equalTo(targetView).offset(margin.top)
    make.left.equalTo(targetView).offset(margin.left)
}

```
and I call `cornerTL.snp.updateConstraints` only once!

**Comment by robertjpayne** [comment]
@tcurdt ah yea so the engine is basically seeing the first constraint and failing the second ( the first is entirely new the second is new but there is now 1 constraint in the list ).

I'll probably tweak this internally so if you call `updateConstraints` and there are 0 constraints it actually runs `makeConstraints` instead.

**Comment by robertjpayne** [comment]
@tcurdt fixed on head of feature/0.40.0 branch!

**Comment by tcurdt** [comment]
Argh! Shouldn't I actually be calling `makeConstraints` all the time?

**Comment by robertjpayne** [comment]
@tcurdt you should:

* Use `makeConstraints` where code executes exactly once OR you need to add more constraints to existing ones
* Use `remakeConstraints` where you have new attribute combinations (e.g. changing left = right to left = left ) and want to throw away all existing constraints and make new ones
* Use `updateConstraints` where  you have identical attribute combinations and just want to update the constant value (e.g. offset(10) to offset(20)

`updateConstraints` and `remakeConstraints` both act like `makeConstraints` if the from view has no existing constraints created from SnapKit.

**Comment by tcurdt** [comment]
Yeah, so it was partly my fault. Thanks for all the quick fixing.
Much appreciated in this horrifying upgrade to swift3 :)

**Comment by robertjpayne** [comment]
@tcurdt no problem, there are some changes specifically `updateConstraints` is less forgiving than it was on 0.22.0 but that's mostly to avoid headache moving forward.

In most cases updateConstraints is only good for seldomly updating constants, if you are going against a gesture or scroll event you really want to store a reference to the Constraint and use it's update APIs.

**Comment by GPF253904828** [comment]
v5.0.1 
![1451596172924_ pic_hd](https://user-images.githubusercontent.com/8220675/89002920-ead5ea00-d330-11ea-8283-656ce2e6af9c.jpg)
![1441596172892_ pic_hd](https://user-images.githubusercontent.com/8220675/89002929-f1646180-d330-11ea-8040-2e93a45c4cde.jpg)

**Comment by fukemy** [comment]
@GPF253904828 You are not updating the constant; you are trying to assign the constraint to a new anchor
